File: 2017/12/p2.py
# Python 3.6.1
# Requires: networkx 1.11

import re

import networkx
from networkx.algorithms import dfs_preorder_nodes as dfs_pn


def get_input():
    with open('input.txt', 'r') as f:
        return [line for line in f]

def build_graph(connections):
    graph = networkx.Graph()
    r = re.compile(r'(\d{1,4}) <-> ((?:\d{1,4},? ?)+)')

    for conn in connections:
        match = r.match(conn)
        source = int(match.group(1))
        dests = [int(d) for d in match.group(2).split(', ')]
        for dest in dests:
            graph.add_edge(source, dest)

    return graph

def main():
    graph = build_graph(get_input())

    remaining = list(dfs_pn(graph))
    count = 0
    while True:
        target = None
        for node in remaining:
            if node in remaining:
                target = node
                break
        remaining = [i for i in remaining if i not in dfs_pn(graph, target)]
        count += 1

        if len(remaining) == 0:
            break

    print(count)


if __name__ == '__main__':
    main()

File: 2016/03/p2.js
// Node 11.2.0

fs = require("fs");

const input = fs.readFileSync("input.txt", "utf-8").split("\n");

function validTriangle({ a, b, c }) {
  return a + b > c && b + c > a && c + a > b;
}

const tris = [];

let tri1 = {};
let tri2 = {};
let tri3 = {};

let step = 0;
let count = 0;

for (let line of input) {
  step++;

  let m = line.match(/(.{5})(.{5})(.{5})/);
  let t1 = parseInt(m[1]);
  let t2 = parseInt(m[2]);
  let t3 = parseInt(m[3]);

  if (step % 3 === 1) {
    tri1.a = t1;
    tri2.a = t2;
    tri3.a = t3;
  } else if (step % 3 === 2) {
    tri1.b = t1;
    tri2.b = t2;
    tri3.b = t3;
  } else if (step % 3 === 0) {
    tri1.c = t1;
    tri2.c = t2;
    tri3.c = t3;

    tris.push(tri1, tri2, tri3);
    tri1 = {};
    tri2 = {};
    tri3 = {};
  }
}

for (let tri of tris) if (validTriangle(tri)) count++;

console.log(count);
